enhancements/authentication/pod-security-admission-autolabeling.md

```diff
@@ -0,0 +1,290 @@
+restriction levels.
+
+The SCC permissions to PSa levels transformation gets explained at the end of this section.
+
+#### Pod Security Admission Introduction
+
+Pod Security admission validates pods' security and distinguishes three different levels:
+- privileged - most privileged mode, anything is allowed
+- baseline   -  minimally restrictive policy which prevents known privilege escalations
+- restricted - heavily restricted policy, following current Pod hardening best practices
+
+By default, there is a cluster-global configuration which enforces the configured policies
+on pods and known workloads.
+
+It is possible to override the cluster-global policy enforce configuration on a per-namespace
+basis by using the `pod-security.kubernetes.io/enforce` label on given namespaces. It is also
+possible to exampt certain users, namespace and runtime classes from the admission completely.
+
+#### PSa Label Synchronization Controller
+
+This section describes a controller that is going to be a part of the cluster-policy-controller.
+
+In order to keep SCCs working alongside restricted cluster-wide PSa level, there must be
+a mechanism that would synchronize the PSa policy level for a namespace that workload runs
+in so that the workload is allowed by both SCC and PSa admissions. For that purpose, a PSa
+Label Synchronization Controller (hereinafter the Controller) shall be built.
+
+The Controller watches changes of:
+- **namespaces**
+  - manual changes to the interesting labels of the namespaces that were
+    opted-in to label synchronization should be reverted
+- all known **workload types** in controlled namespaces
+  - new pods may be getting rejected by PSa for some time before the label
+    synchronization happens
+  - the controller should not allow more relaxed PSa admission level than
+    what SCCs allow
+- **roles** and **rolebindings** (and their **cluster variants**)
+  - to be able to properly assess SA capabilities in controlled namespaces
+- **SCCs**
+  - to properly assess restrictive PSa levels per affected namespaces
+
+The generic control loop of the Controller performs the following actions:
+1. list all controlled (opted-in by a label) namespaces
+2. for each namespace:
+    1. get all instances of known workload types
+    1. associate SCCs to the workloads based on the SAs running them, create a
+       SCC <-> []SA associations
+    1. order observed SCCs most-privileged first (according to PSa privilege levels
+       they map to)
+    1. evaluate the namespace's PSa privilege level based on the most privileged
+       SCC that is allowed to be used by its SA
+    1. modify the namespace's PSa labels based on the enforce level obsedved in the
+       previous step
+
+There is another control loop reacting only to changes of namespaces which performs
+all the steps in 2.
+
+#### Label Synchronization Opting-in
+
+The Controller described in [PSa Label Synchronization Controller](#psa-label-synchronization-controller)
```

Comment:
This sentence doesn't make sense to me, if the PSa controller is opt-in it means you voluntarily sign up for that. So that last part `if they wanted to stick with just SCCs` is not clear to me. If I want to stick with SCC I don't sign up, right?

Reply:
If you don't sign yourself up, you're going to be limited by the cluster-wide pod security admission.

The controller adjusts the registered namespaces so that the PSa level matches the permissions of the SCCs required to run workloads of the said namespace. That means that if you don't register, PSa might block your workloads from running even though you have the proper SCC permissions.

Reply:
I've rewritten this section a bit, hopefully it explains the situation a bit better.